File: 207-CourseSchedule/207-CourseSchedule.py
# Last updated: 8/13/2025, 8:21:15 PM
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = self.calculateIndegree(numCourses , prerequisites)
        neighbours = self.calculateNeighbours(numCourses , prerequisites)
        visited = [False]*numCourses
        
        my_queue = []
        for i in range(len(indegree)) : 
            if indegree[i] == 0 :
                my_queue.append(i)
        
        while len(my_queue) > 0 : 
            popped = my_queue.pop(0)
            
            visited[popped] = True
            neighbour = neighbours[popped]

            for n in neighbour : 
                indegree[n]-=1
                if indegree[n] == 0 : 
                    my_queue.append(n)
        
        for i in range(len(visited)) : 
            if visited[i] == False : 
                return False
        
        return True
        


    def calculateIndegree(self , numCourses : int , prerequisites : List[List[int]]) -> [int] : 

        indegree = [0]*numCourses

        for u , v in prerequisites : 
            indegree[u] +=1
        
        return indegree
    
    def calculateNeighbours(self , numCourses : int , prerequisites : List[List[int]]) -> List[List[int]] : 

        neighbours = [None]*numCourses

        for i in range(numCourses) : 
            neighbours[i] = []
        
        for u , v in prerequisites : 
            neighbours[v].append(u)
        
        return neighbours

File: 207-CourseSchedule/207-CourseSchedule.java
// Last updated: 8/13/2025, 8:21:22 PM
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList();
        for(int i=0;i<numCourses;i++){
            adj.add(new ArrayList());
        }
        for(int[] i : prerequisites){
            int u = i[0];
            int v = i[1];

            adj.get(u).add(v);
        }
        int[] indegree = new int[numCourses];
        for(int i=0;i<numCourses;i++){
            for(int v : adj.get(i)){
                indegree[v]++;
            }
        }
        Queue<Integer> q = new LinkedList();
        for(int i=0;i<numCourses;i++){
            if(indegree[i]==0){
                q.add(i);
            }
        }
        if(q.size()==0)return false;
        while(!q.isEmpty()){
            int rem = q.remove();
            for(int v:adj.get(rem)){
                indegree[v]--;
                if(indegree[v]==0)q.add(v);
            }
        }
        for(int i=0;i<numCourses;i++){
            if(indegree[i]>0){
                return false;
            }
        }
        return true;
    }
}